### PR TITLE
fix/refactor(network::raw): rename raw::ReceivedMessage and accept message types as an arg to send_*()

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
This PR changes things so that instead of calling `send_ping()` and `send_state_part_request()` to send different types of messages, we have `send_message()` and `send_routed_message()` that send direct or routed messages to the peer. This will make it easier to rewrite/add to the code in `tools/mock-node` to use a `raw::Connection` instead of mocking at the actix actor level, because we'll be able to do something like:

```
let queue: Vec<raw::Message> = ...;

match queue[0] {
  raw::Message::Direct(msg) => conn.send_message(msg),
  raw::Message::Routed(msg) => conn.send_routed_message(msg, target, ttl),
};
```

Instead of more cumbersome code that would be required at the moment, especially after adding more message types like `Block`, `BlockRequest`, etc.

This PR also adds a `RouteBackCache` to `raw::Connection`, so that we can tell if routed messages we receive are meant for us. Otherwise, as the code exists now, the `raw::Connection` will pass you any routed messages it receives as if they're meant for you, but they might not be.

### CODE REVIEW

File: chain/network/src/raw/connection.rs
Comment (saketh-are) [inline chain/network/src/raw/connection.rs:1]: The comment describing the overall purpose of this file looks to be outdated (for example, it says "Currently the only message you can send is a Ping message").

**Review by saketh-are — APPROVED** [review]
Some noob questions:
- What do the ping and state-parts tools do?
- Is `raw::Connection` used in any other context? `git grep "use.*raw.*Connection"` seems to indicate no. Based on your PR description I guess the plan is to use it in `mock-node` in the future?
- In what situation (if any) would a user running such a tool receive a RoutedMessage not meant for them?

File: chain/network/src/raw/connection.rs [hunk 1]
```diff
@@ -29,47 +30,56 @@ pub struct Connection {
     stream: TcpStream,
     buf: BytesMut,
     recv_timeout: Duration,
+    // this is used to keep track of routed messages we've sent so that when we get a reply
+    // that references one of our previously sent messages, we can determine that the message is for us
+    route_cache: RouteBackCache,
```
[the remark this answers is not in the document]
Comment (marcelo-gonzalez) [reply chain/network/src/raw/connection.rs:35]: yeah good point

**Comment by marcelo-gonzalez** [comment]
> Some noob questions:
> 
> * What do the ping and state-parts tools do?
> * Is `raw::Connection` used in any other context? `git grep "use.*raw.*Connection"` seems to indicate no. Based on your PR description I guess the plan is to use it in `mock-node` in the future?
> * In what situation (if any) would a user running such a tool receive a RoutedMessage not meant for them?

Yeah at the moment it's just the `ping` and `state-parts` commands that use it. The ping command is kind of helpful for checking what the latencies look like when we send routed messages to the different validators, and the [metrics it exports](https://github.com/near/nearcore/blob/0c673d4470307236aa8b3a17d51822c6361eb7a3/tools/ping/src/metrics.rs) can actually be pretty helpful as an indicator if something is going wrong with the network somewhere. I'm not as familiar with what the state-parts tool is used for, but probably in the ongoing work to redo state sync

As far as receiving a RoutedMessage not meant for us, it seems unlikely at the moment (I don't think I've ever observed that using the ping command), but the old code just feels a bit unclean in that it doesn't event attempt to check that the message is for us. So it's a cleanliness/correctness thing that will usually not matter I guess

**Pull request merged**